src/main/java/core/db/AbstractTable.java

```diff
@@ -3,74 +3,168 @@
-	
-	/** tableName**/
+
+	/**
+	 * tableName
+	 **/
 	private String tableName;
-	
-	/** describes a tableColumn (name, datatype, nullable ..) **/
+
+	/**
+	 * describes a tableColumn (name, datatype, nullable ..)
+	 **/
 	protected ColumnDescriptor[] columns;
-	
-	/** Database connection **/
-	protected  JDBCAdapter  adapter;
-	
+
+	/**
+	 * Database connection
+	 **/
+	protected JDBCAdapter adapter;
+
 	/**
 	 * constructor
+	 *
 	 * @param tableName String table name
 	 */
-	public AbstractTable(String tableName,JDBCAdapter  adapter){
+	public AbstractTable(String tableName, JDBCAdapter adapter) {
 		this.tableName = tableName;
 		this.adapter = adapter;
 		initColumns();
 	}
-	
-	protected String getTableType() {return"CACHED";}
-	
+
+	protected String getTableType() {
+		return "CACHED";
+	}
+
 	protected abstract void initColumns();
-	
-	
-	protected String getTableName(){
+
+
+	protected String getTableName() {
 		return tableName;
 	}
-	protected void setColumns(ColumnDescriptor[] columns){
+
+	protected void setColumns(ColumnDescriptor[] columns) {
 		this.columns = columns;
 	}
-	
-	protected ColumnDescriptor[] getColumns(){
+
+	protected ColumnDescriptor[] getColumns() {
 		return columns;
 	}
-	
-	protected String[] getCreateIndexStatement(){
+
+	protected String[] getCreateIndexStatement() {
 		return new String[0];
 	}
-	
-	protected String[] getConstraintStatements(){
+
+	protected String[] getConstraintStatements() {
 		return new String[0];
 	}
-	
-	protected int delete(String[] whereColumns, String[] whereValues) {
-		
-		final StringBuilder sql = new StringBuilder("DELETE FROM ");
-		sql.append(getTableName());
 
-		//Where bedingungen beachten
-		if ((whereValues != null) && (whereColumns != null) && (whereColumns.length == whereValues.length) && (whereValues.length > 0)) {
-			sql.append(" WHERE ").append(whereColumns[0]).append(" = ").append(whereValues[0]);
+	// Prepared Statements
+	// Insert
+	public static class PreparedInsertStatementBuilder extends DBManager.PreparedStatementBuilder {
+		public PreparedInsertStatementBuilder(AbstractTable table) {
+			super(table.createInsertStatement());
+		}
+	}
+	protected PreparedInsertStatementBuilder preparedInsertStatementBuilder;
+	protected PreparedInsertStatementBuilder createPreparedInsertStatementBuilder(){
+		return new PreparedInsertStatementBuilder(this);
+	}
+	private String createInsertStatement() {
+		var ret = new StringBuilder("INSERT INTO ");
+		ret.append(getTableName());
+		var valuePlaceholders = new StringBuilder(" VALUES ");
+		var sep = "(";
+		for (var c : columns) {
+			valuePlaceholders.append(sep).append("?");
+			ret.append(sep).append(c.getColumnName());
+			sep = ",";
+		}
+		ret.append(")").append(valuePlaceholders).append(")");
+		return ret.toString();
+	}
+	protected int executePreparedInsert(Object ... values){
+		if ( preparedInsertStatementBuilder==null){
+			preparedInsertStatementBuilder=createPreparedInsertStatementBuilder();
```

Comment:
I must admit I don't quite understand the added complexity of using `PreparedStatementBuilder`, which AFAICT don't do much (other than creating the `PreparedStatement` and caching it).  I thought this could have been because they get overridden in subclasses, but that doesn't seem to be the case either.  Would it not be simpler to call directly `JDBCAdapater#createPreparedStatement`?  Am I missing something?

Also, I am worried that those statements are left open without seemingly being closed, which could cause some DB issues.   Is that the case?  A while back I had started some work to ensure all statements and result sets were getting closed properly, but hadn't had the chance to finish...  Except now it looks like it would be straightforward, as all the executions seem to be localised to this class.

Reply:
@tychobrailleur  thank you very much for your input!!!

concerning Builder: I introduced the builder when it became annoying for me to type the same three lines over and over again. later on i found it good to create specialized Builder with standard sqls for inserting, updating and deleting in abstractTable. Overriding is delegated to the standard Create-Builder methods provided in AbstractTable.

concerning Close: I hope they don't need any closing, since their usage is reusing. If I'm wrong, we now have a central place to handle this, when ho is terminated.

Reply:
Ok, it would be good to keep an eye on all these cached preparedstatements, and see if that causes issues further down the line.